# Contextual retrieval and token witnesses

This companion examines context without requiring the original name words.
It audits full-amplitude queries, without support-floor or atom reduction.
The saved experiment uses both GPT-2 sparse surrogate families at blocks
8 and 11. It samples three eligible rows for each of four queries per
condition, retaining every sampled case, including semantic mismatches.

The corpus objects are cached 128-position OpenWebText blocks, not uniquely
identified original documents. See the protocol for source rows, lexical
exclusions, sample seed, precision, and limitations.


Panel headings use backbone size, surrogate mechanism, zero-based block,
and activation site. Exact releases and checkpoint IDs are recorded in
`data/activation_studies/tablecontexts_v1/identities.json`. The separate
`commoncontexts_codexgen.ipynb` demonstrates an uncalibrated five-document
meet; it does not replace this notebook's query protocol.


In [1]:
from pathlib import Path
import json
import os
import sys

root = next(p for p in [Path.cwd(), *Path.cwd().parents]
            if (p / "src/lattmc/contextstudy").is_dir())
os.environ["MY_PAPERS_REPOSITORY"] = str(root)
sys.path.insert(0, str(root / "src"))
output = root / "data/activation_studies/context_v1"
print("Results:", output)


Results: /Users/ltsinadze/git/my_papers/data/activation_studies/context_v1


## Independent saved-array audit

Verify seeded selection, word exclusions, every coordinate witness,
exact fresh membership, unchanged token multisets, diagnostic threshold
crossings, and suffix-control equality. No neural inference occurs here.


In [2]:
from lattmc.contextstudy.audit_codexgen import main as audit

result = audit()
assert result["counts"]["rows"] == 48
assert result["counts"]["fresh_members"] == 48


{
  "rows": 48,
  "fresh_members": 48,
  "prefix_trials": 144,
  "changed_prefix_trials": 119,
  "material_changes": 119,
  "witness_threshold_losses": 95,
  "suffix_controls": 48
}
Maximum suffix-control error: 0.0


In [3]:
import pandas as pd
from IPython.display import display

display(pd.DataFrame(result["conditions"]))


,kind,layer,case,extent,eligible,sampled,prefix,changes,losses
0,sae,8,nyc,4935,4033,3,3,3,3
1,sae,8,rio,7,6,3,9,9,6
2,sae,8,animals,81,50,3,9,9,4
3,sae,8,sports,148,137,3,9,9,6
4,sae,11,nyc,925,730,3,7,7,7
5,sae,11,rio,20,8,3,9,9,6
6,sae,11,animals,1287,1219,3,7,7,6
7,sae,11,sports,21,16,3,9,9,9
8,tc,8,nyc,4184,3512,3,0,0,0
9,tc,8,rio,8396,8366,3,9,9,9


## Complete query inspection and the yellow convention

Retrieval uses the pooled item vector. A single token need not satisfy
all coordinates. The shared companion now checks every valid token
against the same scaled query used for item membership. S means at least
one whole-query token; yellow D means distributed satisfaction; R means
rejection. Rose marks any coordinate witness and bold marks whole-query
tokens. Yellow is not a semantic correctness annotation.

The appendix shows the entire cached span. The main paper uses one fixed
64-token window rule. The expandable ledger reports every coordinate's
requirement and maximal measured token activation. No score or selection
protocol is changed by this presentation update.


In [4]:
from IPython.display import HTML, display
from lattmc.contextstudy.witnessaudit_codexgen import main as witness_audit
from lattmc.contextstudy.witnessrender_codexgen import html_gallery, records

witness_audit()
rows = [r for r in records() if r['group'] == 'context']
display(HTML(html_gallery(rows)))

{'status': 'passed', 'rows': 72, 'coordinates': 306, 'sparse_dense_library_comparisons': 192, 'boundary_cases': 6, 'counts': {'context': {'S': 15, 'D': 33, 'R': 0, 'N': 0}, 'family': {'S': 15, 'D': 8, 'R': 1, 'N': 0}}, 'source_sha256': 'ff0401614c355fbfee42d37683ffccf9843c246398d7342d41f71fb8f858be15', 'legacy_rows': 22, 'unresolved_historical_rows': 4}


Coordinate,Requirement,Maximal token,Activation,Pass
7000,1.4353956,9,15.1452522,True
7013,0.402003616,6,1.41461587,True
7055,9.39362717,6,12.2265368,True
22057,3.08952475,9,10.4546089,True
Coordinate,Requirement,Maximal token,Activation,Pass
7000,1.4353956,3,2.04199004,True
7013,0.402003616,4,3.86545038,True
7055,9.39362717,4,11.3361073,True
22057,3.08952475,3,4.96906805,True
Coordinate,Requirement,Maximal token,Activation,Pass


## Interpretation and rerun commands

The regional and baboon passages illustrate related contexts without the
source name words. The medical `catheter` passage illustrates why word
exclusion alone is insufficient. The source Cat occurs in a film title;
it is not an independently assigned biological label.

All 119 nonidentity prefix permutations change the selected activation;
95 lower that witness below the query threshold. All 48 suffix controls
are unchanged. These results demonstrate prefix-order sensitivity at
selected positions, not semantic accuracy or a traced MLP-only circuit.
They do not imply that a whole-row max-pooled query stops retrieving a row.

Neural inference was executed from the source module. To repeat it with
the existing project uv environment and cached inputs, run from the root:

```sh
PYTHONPATH=src uv run --locked python -m lattmc.contextstudy.run_codexgen
PYTHONPATH=src uv run --locked python -m lattmc.contextstudy.audit_codexgen
```
